Sources/SwiftFusion/Inference/BearingRangeFactor.swift

```diff
@@ -0,0 +1,193 @@
+// Copyright 2019 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import TensorFlow
+//
+///// Protocol for indicating a struct is range-capable
+//public protocol HasRange: Differentiable {
+//  associatedtype A: Differentiable
+//  associatedtype B: Differentiable where B.TangentVector: EuclideanVectorSpace & VectorConvertible & AdditiveArithmetic
+//  @differentiable
+//  static func range(_ from: A, _ to: B) -> Double
+//}
+//
+//public struct PoseToPointRange2D: HasRange {
+//  @differentiable
+//  public static func range(_ from: Pose2, _ to: Vector2) -> Double {
+//    (to - from.t).norm
+//  }
+//}
+//
+//public struct PointToPointRange2D: HasRange {
+//  @differentiable
+//  public static func range(_ from: Vector2, _ to: Vector2) -> Double {
+//    (to - from).norm
+//  }
+//}
+//
+///// Protocol for indicating a struct is bearing-capable
+//public protocol HasBearing: Differentiable {
+//  associatedtype A: LieGroup where A.TangentVector: VectorConvertible
+//  associatedtype B: Differentiable where B.TangentVector: EuclideanVectorSpace & VectorConvertible & AdditiveArithmetic
+//  associatedtype C: EuclideanVectorSpace & Differentiable & AdditiveArithmetic & VectorProtocol & TangentStandardBasis
+//  @differentiable
+//  static func bearing(_ from: A, _ to: B) -> C
+//}
+//
+//public struct PoseToPointBearing2D: HasBearing {
+//  @differentiable
+//  public static func bearing(_ from: Pose2, _ to: Vector2) -> Vector1 {
+//    let dx = (to - from.t)
+//    return Vector1(between(Rot2(c: dx.x / dx.norm, s: dx.y / dx.norm), from.rot).theta)
+//  }
+//}
+//
+//public struct BearingRangeError<BE: TangentStandardBasis, RE: TangentStandardBasis>:
```

Comment:
Is `BearingRangeError` your solution to the problem that "It is not trivial to combine two fixed-size vectors into one that is of a higher dimension" ?

I think this is the right solution! The `BearingRangeError` _is_ the combined vector of a higher dimension. Do you  think it's a good solution? Is there anything that would be missing if you could get this working?

Reply:
The problem here is that I cannot make this conform to `VectorConvertible`,  as it appears to me that there is no way to get the dimension of the two component types: for example, 2d bearing + range is (1 + 1), 3d bearing + range is (3 + 1) and possibly others...

Reply:
Ah, I see. This should help: https://github.com/borglab/SwiftFusion/pull/62 !

Reply:
Thanks Marc! One (hopefully) last problem: How to make differentiable slices of an array?
i.e. this won't work (not differentiable):

![image](https://user-images.githubusercontent.com/5358374/82238954-0cb0fa80-9906-11ea-8117-70bce3440236.png)
```swift
public struct BearingRangeError<BE: TangentStandardBasis & VectorConvertible & FixedDimensionVector>:
  Differentiable & KeyPathIterable & TangentStandardBasis & VectorConvertible {
  @differentiable
  public init(_ vector: Vector) {
    precondition(vector.dimension == BE.dimension + 1)
    bearing = BE(Vector(Array(vector.scalars[0..<BE.dimension])))
    range = vector.scalars[BE.dimension]
  }
  
  @differentiable
  public init(bearing: BE, range: Double) {
    self.bearing = bearing
    self.range = range
  }

  public var vector: Vector {
    get {
      Vector(bearing.vector.scalars + [range])
    }
  }

  public var bearing: BE
  public var range: Double
}
```

Reply:
Ah, we should definitely update the stdlib so that array slices are differentiable without any extra work.

In the meantime, here's a method that should help:
```swift
extension Array where Element: Differentiable {
  struct Partitioned: Differentiable {
    var a: [Element]
    var b: [Element]
  }

  @differentiable(wrt: self)
  func differentiablePartition(_ index: Int) -> Partitioned {
    return Partitioned(a: Array(self[0..<index]), b: Array(self[index..<endIndex]))
  }

  @derivative(of: differentiablePartition)
  func vjpDifferentiablePartition(_ index: Int) -> (value: Partitioned, pullback: (Partitioned.TangentVector) -> TangentVector) {
    let partitioned = differentiablePartition(index)
    let aCount = partitioned.a.count
    let bCount = partitioned.b.count
    return (
      partitioned,
      {
        let aTv = $0.a.base.count > 0 ? $0.a.base : Array<Element.TangentVector>(repeating: .zero, count: aCount)
        let bTv = $0.b.base.count > 0 ? $0.b.base : Array<Element.TangentVector>(repeating: .zero, count: bCount)
        return TangentVector(aTv + bTv)
      }
    )
  }
}
```

I wrote this super quickly and haven't had time to test it very thoroughly, and now I have to go to a meeting. But I did some simple tests and it seems correct.